tv_apps/tv_shared/js/card_manager.js

```diff
@@ -206,51 +193,62 @@
       }
     },
 
+    _initCardStoreIfNeeded: function cm_initCardStore() {
+      if (!this._cardStore) {
+        this._cardStore =
+          new CardStore(this._mode, this._manifestURLOfCardStore);
+        this._cardStore.on('change', this._onCardStoreChange.bind(this));
+      }
+    },
+
+    // XXX: DO NOT call this directly except for _reloadCardList.
+    // Because it is not protected by AsyncSemaphore
+    _loadCardListFromCardStore:
+      function cm_loadCardListFromCardStore(cardListFromCardStore) {
+        var that = this;
+        cardListFromCardStore.forEach(function(cardEntry) {
+          var found = that.findCardFromCardList({cardEntry: cardEntry});
+          if (!found) {
+            var card = that._deserializeCardEntry(cardEntry);
+            // The cards inside of folder are not saved nested in
+            // datastore 'cardList'. But we explicit save them under key
+            // of folderId. So we need to retrieve them by their folderId
+            // and put them back to folders where they belong.
+            if (card instanceof Folder &&
+                card.state === Folder.STATES.DESERIALIZING) {
+              // to load content of folder
+              that._cardStore.getData(card.folderId).then(
+                function(innerCardList) {
+                  innerCardList.forEach(function(innerCardEntry) {
+                    card.cardsInFolder.push(
+                      that._deserializeCardEntry(innerCardEntry));
+                  });
+                });
+            }
+            that._cardList.push(card);
+          }
+        });
+      },
+
     _reloadCardList: function cm_loadCardList() {
       var that = this;
       return this._getPipedPromise('_reloadCardList',
         function(resolve, reject) {
-          // initialize card store if needed
-          if (!that._cardStore) {
-            that._cardStore =
-              new CardStore(that._mode, that._manifestURLOfCardStore);
-            that._cardStore.on('change', that._onCardStoreChange.bind(that));
+          that._asyncSemaphore.v();
+          that._initCardStoreIfNeeded();
+          resolve(that._cardStore.getData('cardList'));
+        }).then(function(cardListFromCardStore) {
+          if (cardListFromCardStore) {
+            that._loadCardListFromCardStore(cardListFromCardStore);
```

Comment:
Since we don't have Folder now, this wouldn't be a problem. But if we land it as is I'd suggest add some comment here like `XXX: Change _loadCardListFromCardStore as returning a promise`.